### PR TITLE
[enhancement](cloud) batching get visible version from MetaService

### PR DESCRIPTION
**Pull request by freemandealer**
Get visible versions one by one from MetaService is costly. Batching them into one RPC will not only reduce the workload of RPC service but also reduce the lag.

### CODE REVIEW

**Comment by freemandealer** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java [hunk 1]
```diff
@@ -253,16 +257,28 @@ private boolean tablesOrDataChanged(Env env, SqlCacheContext sqlCacheContext) {
                 return true;
             }
             OlapTable olapTable = (OlapTable) tableIf;
-            long currentTableTime = olapTable.getVisibleVersionTime();
-            long cacheTableTime = scanTable.latestTimestamp;
+            //long currentTableTime = olapTable.getVisibleVersionTime();
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:260]: revmove unused code

Comment (freemandealer) [reply fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:260]: ok, will do after all these p0 cases passes.

File: gensrc/proto/cloud.proto [hunk 2]
```diff
@@ -662,6 +663,7 @@ message CommitTxnResponse {
     repeated int64 partition_ids = 4;
     repeated int64 versions = 5;
     repeated TableStatsPB table_stats = 6;
+    optional int64 version_update_time = 7;
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:666]: What is this for? Why is version_update_time related to RecycleTxnPB

File: gensrc/proto/cloud.proto [hunk 3]
```diff
@@ -759,6 +761,8 @@ message GetVersionResponse {
     repeated int64 table_ids = 4;
     repeated int64 partition_ids = 5;
     repeated int64 versions = 6;
+    repeated int64 version_update_times = 7;
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:764]: what's the single and multiple update_time for?

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 4]
```diff
@@ -1043,10 +1043,14 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
     }
 
     // Save versions
+    int64_t version_update_time =
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1046]: add time unit as suffix 
e.g. `version_update_time_ms`

pls. check all the other occurrences

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java [hunk 5]
```diff
@@ -253,16 +257,28 @@ private boolean tablesOrDataChanged(Env env, SqlCacheContext sqlCacheContext) {
...
+                List<CloudPartition> partitions = partitionIds.stream()
+                        .sorted()
+                        .map(olapTable::getPartition)
+                        .map(partition -> (CloudPartition) partition)
+                        .collect(Collectors.toList());
+                try {
+                    CloudPartition.getSnapshotVisibleVersion(partitions);
+                } catch (RpcException e) {
+                    throw new RuntimeException(e);
+                }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:280]: add comment to explain what this code block use for

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 6]
```diff
@@ -697,15 +700,29 @@ private CacheTable buildCacheTableForOlapScanNode(OlapScanNode node) {
...
+                    .sorted()
+                    .map(olapTable::getPartition)
+                    .map(partition -> (CloudPartition) partition)
+                    .collect(Collectors.toList());
+            try {
+                CloudPartition.getSnapshotVisibleVersion(partitions);
+            } catch (RpcException e) {
+                throw new RuntimeException(e);
+            }
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:718]: add comment to explain it

File: gensrc/proto/cloud.proto (hunk 2 above)
Comment (freemandealer) [reply gensrc/proto/cloud.proto:666]: They are for batch and single respectively. What makes you think the relation ever existed?

File: gensrc/proto/cloud.proto (hunk 3 above)
Comment (freemandealer) [reply gensrc/proto/cloud.proto:764]: ditto

File: cloud/src/meta-service/meta_service_txn.cpp (hunk 4 above)
Comment (freemandealer) [reply cloud/src/meta-service/meta_service_txn.cpp:1046]: sure, will do

File: fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java (hunk 5 above)
Comment (freemandealer) [reply fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:280]: sure, will do

File: cloud/src/meta-service/meta_service.cpp [hunk 7]
```diff
@@ -261,6 +261,7 @@ void MetaServiceImpl::get_version(::google::protobuf::RpcController* controller,
                 return;
             }
             response->set_version(version);
+            // TODO response->set_version_update_time_ms();
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service.cpp:264]:
```suggestion
```

File: cloud/src/meta-service/meta_service.cpp [hunk 8]
```diff
@@ -381,6 +383,7 @@ void MetaServiceImpl::batch_get_version(::google::protobuf::RpcController* contr
                         break;
                     }
                     response->add_versions(version);
+                    // TODO response->add_version_update_times_ms();
```
Comment (w41ter) [inline cloud/src/meta-service/meta_service.cpp:386]:
```suggestion
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java [hunk 9]
```diff
@@ -697,15 +700,29 @@ private CacheTable buildCacheTableForOlapScanNode(OlapScanNode node) {
         CatalogIf catalog = database.getCatalog();
         ScanTable scanTable = new ScanTable(
                 new FullTableName(catalog.getName(), database.getFullName(), olapTable.getName()),
-                olapTable.getVisibleVersionTime(), olapTable.getVisibleVersion());
+                olapTable.getVisibleVersion());
         scanTables.add(scanTable);
+        if (Config.isCloudMode()) {
+            // get version of related partitions in batch
```
Comment (w41ter) [inline fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:706]: How about extracting this operation into `OlapTable`, just like `OlapTable.selectNonEmptyPartitionIds`?

**Comment by freemandealer** [comment]
run buildall

**Review by gavinchou — COMMENTED** [review]
LGTM\

File: fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java (hunk 9 above)
Comment (freemandealer) [reply fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:706]: Ops! On second thought, I think can give it a try.

**Review by w41ter — DISMISSED** [review]

**Comment by freemandealer** [comment]
run buildall

**Review by dataroaring — DISMISSED** [review]
LGTM

**Comment by freemandealer** [comment]
run buildall

**Review by 924060929 — DISMISSED** [review]
LGTM

**Comment by freemandealer** [comment]
run cloud_p1

**Review by w41ter — DISMISSED** [review]

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run buildall

File: gensrc/proto/cloud.proto [hunk 10]
```diff
@@ -818,6 +820,8 @@ message GetVersionResponse {
     repeated int64 table_ids = 4;
     repeated int64 partition_ids = 5;
     repeated int64 versions = 6;
+    repeated int64 version_update_times_ms = 7; // for batch
+    optional int64 version_update_time_ms = 8; // for single
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:824]: Consider using only `version_update_times_ms`

**Comment by freemandealer** [comment]
run buildall

**Comment by freemandealer** [comment]
run p0

**Comment by freemandealer** [comment]
run p0

**Comment by freemandealer** [comment]
run p0

**Review by gavinchou — APPROVED** [review]

**Review by Yukang-Lian — APPROVED** [review]

**Review by sollhui — APPROVED** [review]
LGTM

**Pull request merged**